fp-core/src/hkt.rs

```diff
@@ -1,13 +1,21 @@
-pub trait HKT<A, B> {
-    type URI;
-    type Target;
+pub trait HKT<U> {
+    type Current; // The current contained type
+    type Target; // The next type (how to generically fill in U)
 }
 
-impl<A, B> HKT<A, B> for Option<A> {
-    type URI = Self;
-    type Target = Option<B>;
+macro_rules! derive_hkt {
```

Comment:
hmmmmmm,  can we have a chat about this?

Reply:
Sure...

The basic point is that any hkt with 1 parameter has this HKT instance: it contains `T` and can be converted to contain `U`.

Reply:
Ok after playing around with your new HKT idea, I think the macro idea is amazing but I feel like doing `Self as HKT<B>>::Current` instead of simply using generic `A` might become very difficult to read in the future T_T... I'm already finding it quite verbose. Can we turn back `Self as HKT<B>>::Current` into `A` as before?